### PR TITLE
[Impeller] Correct attachment description for offscreen MSAA resolve.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Two issues:

1. The resolve texture is created from the same impeller attachment as the msaa attachment, so the store mode was getting set to dont care instead of store.

2. the image layout for the resolve attachment should be ePresentSrcKHR, at least from following the guide at https://vulkan-tutorial.com/Multisampling

Fixing both of these locally fixes all of the weird offscreen cursed rendering.

Fixes https://github.com/flutter/flutter/issues/128600

### CODE REVIEW

**Comment by jonahwilliams** [comment]
lol now we get a validation error. I think I just need to correct the layout info.

**Comment by jonahwilliams** [comment]
hilariously I misread the tutorial, but visually things are fixed for me 😮‍💨

**Comment by jonahwilliams** [comment]
So I suspect that rather than fixing this, the error has introduced some pessimization which is causing the synchronization issue to resolve itself.

File: impeller/renderer/backend/vulkan/formats_vk.h [hunk 1]
```diff
@@ -466,7 +467,9 @@ constexpr vk::AttachmentDescription CreateAttachmentDescription(
   switch (kind) {
     case AttachmentKind::kColor:
       vk_attachment.initialLayout = current_layout;
-      vk_attachment.finalLayout = vk::ImageLayout::eColorAttachmentOptimal;
+      vk_attachment.finalLayout =
+          resolve_texture ? vk::ImageLayout::ePresentSrcKHR
+                          : vk::ImageLayout::eColorAttachmentOptimal;
```
Comment (dnfield) [inline impeller/renderer/backend/vulkan/formats_vk.h:472]: Is this the right switch value?

Spec says that eColorAttachmentOptimal "must only" be used as a color ro resolve in a `VkFramebuffer`. 

Perhaps the argument should instead be `bool is_framebuffer` or something? Or perhaps an enum where the enumerated values are something like `kFramebuffer` and `kNonFramebuffer`.

Comment (dnfield) [reply impeller/renderer/backend/vulkan/formats_vk.h:472]: I'm not sure I really understand what `ePresentSrcKHR` is really for. Did you already try `eGeneral`?

File: impeller/renderer/backend/vulkan/render_pass_vk.cc [hunk 2]
```diff
@@ -95,7 +98,7 @@ SharedHandleVK<vk::RenderPass> RenderPassVK::CreateVKRenderPass(
     if (color.resolve_texture) {
       resolve_refs[bind_point] =
           vk::AttachmentReference{static_cast<uint32_t>(attachments.size()),
-                                  vk::ImageLayout::eColorAttachmentOptimal};
+                                  vk::ImageLayout::ePresentSrcKHR};
```
Comment (dnfield) [inline impeller/renderer/backend/vulkan/render_pass_vk.cc:101]: (Ditto here: maybe eGeneral instead?)

File: impeller/renderer/backend/vulkan/render_pass_vk.cc [hunk 3]
```diff
@@ -49,14 +49,17 @@ static vk::AttachmentDescription CreateAttachmentDescription(
 
   if (desc.storage_mode == StorageMode::kDeviceTransient) {
```
Comment (dnfield) [inline impeller/renderer/backend/vulkan/render_pass_vk.cc:50]: Is the real problem that we're setting the storage mode to device transient?

I'm not quite clear on when we should set dontcare or not though.

File: impeller/renderer/backend/vulkan/formats_vk.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/formats_vk.h:472]: I don't really understand either, but it was the value used in https://vulkan-tutorial.com/Multisampling so I just went with it. I'll try general

File: impeller/renderer/backend/vulkan/render_pass_vk.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/render_pass_vk.cc:50]: no that doesn't seem to be it, at least things are already working for me

File: impeller/renderer/backend/vulkan/formats_vk.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/formats_vk.h:472]: eGeneral works! lets see if it validates

File: impeller/renderer/backend/vulkan/formats_vk.h [hunk 4]
```diff
@@ -466,7 +467,9 @@ constexpr vk::AttachmentDescription CreateAttachmentDescription(
   switch (kind) {
     case AttachmentKind::kColor:
       vk_attachment.initialLayout = current_layout;
-      vk_attachment.finalLayout = vk::ImageLayout::eColorAttachmentOptimal;
+      vk_attachment.finalLayout =
+          resolve_texture ? vk::ImageLayout::eGeneral
```
Comment (dnfield) [inline impeller/renderer/backend/vulkan/formats_vk.h:471]: The nitpick here is using "resolve_texture" as the reason to do this.

The not nitpick here is that I think we should go with eGeneral unless we know for sure that this is for the framebuffer. eGeneral should always be safe if I'm understanding the spec, whereas the eColorAttachmentOptimal is only safe if we know we're using this for framebuffer.

**Review by dnfield — APPROVED** [review]

**Pull request merged**